pkg/storage/read_as_of_iterator.go

```diff
@@ -94,42 +100,85 @@ func (f *ReadAsOfIterator) UnsafeValue() []byte {
+// rangekeys.
 func (f *ReadAsOfIterator) RangeBounds() roachpb.Span {
-	panic("not implemented")
+	return roachpb.Span{}
 }
 
-// RangeKeys implements SimpleMVCCIterator.
+// RangeKeys is always empty since this iterator never surfaces rangeKeys.
 func (f *ReadAsOfIterator) RangeKeys() []MVCCRangeKeyValue {
-	panic("not implemented")
+	return []MVCCRangeKeyValue{}
+}
+
+// updateValid updates i.valid and i.err based on the underlying iterator, and
+// returns true if valid.
+func (f *ReadAsOfIterator) updateValid() bool {
+	f.valid, f.err = f.iter.Valid()
+	return f.valid
 }
 
 // advance moves past keys with timestamps later than f.asOf and skips MVCC keys
-// whose latest value (subject to f.asOF) has been deleted. Note that advance
-// moves past keys above asOF before evaluating tombstones, implying the
-// iterator will never call f.iter.NextKey() on a tombstone with a timestamp
-// later than f.asOF.
-func (f *ReadAsOfIterator) advance() bool {
+// whose latest value (subject to f.asOF) has been deleted by a point or range
+// tombstone.
+func (f *ReadAsOfIterator) advance() {
 	for {
-		if ok, err := f.Valid(); err != nil || !ok {
-			// No valid keys found.
-			return false
-		} else if !f.asOf.IsEmpty() && f.asOf.Less(f.iter.UnsafeKey().Timestamp) {
-			// Skip keys above the asOf timestamp.
+		if ok := f.updateValid(); !ok {
+			return
+		}
+
+		if !f.asOf.IsEmpty() && f.asOf.Less(f.iter.UnsafeKey().Timestamp) {
+			// Skip keys above the asOf timestamp, regardless of type of key (e.g. point or range)
+			f.iter.Next()
+		} else if hasPoint, hasRange := f.iter.HasPointAndRange(); !hasPoint && hasRange {
+			// Bare range keys get surfaced before the point key, even though the
+			// point key shadows it; thus, because we can infer range key information
+			// when the point key surfaces, skip over the bare range key, and reason
+			// about shadowed keys at the surfaced point key.
+			//
+			// E.g. Scanning the keys below:
+			//  2  a2
+			//  1  o---o
+			//     a   b
+			//
+			//  would result in two surfaced keys:
+			//   {a-b}@1;
+			//   a2, {a-b}@1
+
 			f.iter.Next()
 		} else if len(f.iter.UnsafeValue()) == 0 {
-			// Skip to the next MVCC key if we find a tombstone.
+			// Skip to the next MVCC key if we find a point tombstone.
+			f.iter.NextKey()
+		} else if !hasRange {
+			// On a valid key without a range key
+			return
+		} else if f.asOfRangeKeyShadows() {
+			// The latest range key, as of system time, shadows the latest point key.
+			// This key is therefore deleted as of system time.
 			f.iter.NextKey()
 		} else {
-			// On a valid key.
-			return true
+			// On a valid key that potentially shadows range key(s)
+			return
 		}
 	}
 }
 
+// asOfRangeKeyShadows returns true if there exists a range key at or below the asOf timestamp
+// that shadows the latest point key
+func (f *ReadAsOfIterator) asOfRangeKeyShadows() (shadows bool) {
+	rangeKeys := f.iter.RangeKeys()
```

Comment:
Btw, this is going to be really expensive, since we're reconstructing the range key stack for every point key. We may want to cache it until `RangeBounds().Key` changes (or wait until #84379 lands).

We should also do some end-to-end benchmarking of these bulk operations before 22.2, since I'm guessing we may see some hefty performance regressions that we'll need to address.

Reply:
Yeah, it'll be rare, but we also want to avoid falling of a performance cliff when it does happen.

Reply:
will do!

Reply:
https://github.com/cockroachdb/cockroach/issues/84714